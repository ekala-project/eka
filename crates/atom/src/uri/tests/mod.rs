use super::*;

const ALIASES: &[&str] = &[
    "gl:repo::atom@^2.0",
    "gl::atom@^2.1",
    "gl:path.with/dot::my-atom@^2",
    "git@github.com:owner/repo::this-atom@^1",
    "git@gh:owner/repo::this-atom@^1",
    "https://example.com:8080/owner/repo::foo@^1",
    "https://foo.com/owner/repo::bar@^1",
    "https://gh:owner/repo::λ@^1",
    "gh:owner/repo::λ@^1",
    "git@gh:owner/repo::λ@^1",
    "pkgs::zlib@^1",
    "pkgs:foo/bar::baz",
    "https://user:password@example.com/my/repo::id@^0.2",
    "user:password@example.com/my/repo::id@^0.2",
    "user@example.com/my/repo::id@^0.2",
    "gh:owner/repo::yep@^1",
    "bb::yep@^1",
    "foo/bar/baz::my-atom",
    "/foo/bar/baz::my-atom",
    // not an alias, but an ssh url without a username
    "my.ssh.com:my/repo::hello",
    "foo@^0.8",
    "::foo",
    "::foo",
];

const ALIASED_URLS: &[&str] = &[
    "gh:foo/bar^master",
    "bb^refs/heads/my-work",
    "https://gl:bar/baz",
    "pkgs^main",
    "git@gh:owner/repo^master",
];

#[test]
fn ref_snapshot() {
    let results: Vec<Ref> = ALIASES.iter().map(|x| (*x).into()).collect();
    insta::assert_yaml_snapshot!(results);
}

#[test]
fn uri_snapshot() -> Result<(), UriError> {
    let results: Result<Vec<Uri>, UriError> = ALIASES.iter().map(|v| v.parse::<Uri>()).collect();
    insta::assert_debug_snapshot!(results?);
    Ok(())
}

#[test]
fn url_snapshot() -> anyhow::Result<()> {
    let parsed: Vec<AliasedUrl> = ALIASED_URLS
        .iter()
        .map(|x| (*x).try_into())
        .collect::<Result<Vec<_>, _>>()?;
    let results: Vec<_> = parsed.iter().map(|x| x.to_string()).collect();
    insta::assert_yaml_snapshot!(results);
    Ok(())
}
