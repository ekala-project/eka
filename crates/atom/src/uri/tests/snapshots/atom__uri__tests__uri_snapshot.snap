---
source: crates/atom/src/uri/tests/mod.rs
assertion_line: 46
expression: results?
---
[
    Uri {
        url: Some(
            Url {
                scheme: Https,
                user: None,
                password: None,
                host: Some(
                    "gitlab.com",
                ),
                serialize_alternative_form: false,
                port: None,
                path: "/repo",
            },
        ),
        tag: AtomTag(
            "atom",
        ),
        version: Some(
            VersionReq {
                comparators: [
                    Comparator {
                        op: Caret,
                        major: 2,
                        minor: Some(
                            0,
                        ),
                        patch: None,
                        pre: Prerelease(""),
                    },
                ],
            },
        ),
    },
    Uri {
        url: Some(
            Url {
                scheme: Https,
                user: None,
                password: None,
                host: Some(
                    "gitlab.com",
                ),
                serialize_alternative_form: false,
                port: None,
                path: "/",
            },
        ),
        tag: AtomTag(
            "atom",
        ),
        version: Some(
            VersionReq {
                comparators: [
                    Comparator {
                        op: Caret,
                        major: 2,
                        minor: Some(
                            1,
                        ),
                        patch: None,
                        pre: Prerelease(""),
                    },
                ],
            },
        ),
    },
    Uri {
        url: Some(
            Url {
                scheme: Https,
                user: None,
                password: None,
                host: Some(
                    "gitlab.com",
                ),
                serialize_alternative_form: false,
                port: None,
                path: "/path.with/dot",
            },
        ),
        tag: AtomTag(
            "my-atom",
        ),
        version: Some(
            VersionReq {
                comparators: [
                    Comparator {
                        op: Caret,
                        major: 2,
                        minor: None,
                        patch: None,
                        pre: Prerelease(""),
                    },
                ],
            },
        ),
    },
    Uri {
        url: Some(
            Url {
                scheme: Ssh,
                user: Some(
                    "git",
                ),
                password: None,
                host: Some(
                    "github.com",
                ),
                serialize_alternative_form: false,
                port: None,
                path: "/owner/repo",
            },
        ),
        tag: AtomTag(
            "this-atom",
        ),
        version: Some(
            VersionReq {
                comparators: [
                    Comparator {
                        op: Caret,
                        major: 1,
                        minor: None,
                        patch: None,
                        pre: Prerelease(""),
                    },
                ],
            },
        ),
    },
    Uri {
        url: Some(
            Url {
                scheme: Ssh,
                user: Some(
                    "git",
                ),
                password: None,
                host: Some(
                    "github.com",
                ),
                serialize_alternative_form: false,
                port: None,
                path: "/owner/repo",
            },
        ),
        tag: AtomTag(
            "this-atom",
        ),
        version: Some(
            VersionReq {
                comparators: [
                    Comparator {
                        op: Caret,
                        major: 1,
                        minor: None,
                        patch: None,
                        pre: Prerelease(""),
                    },
                ],
            },
        ),
    },
    Uri {
        url: Some(
            Url {
                scheme: Https,
                user: None,
                password: None,
                host: Some(
                    "example.com",
                ),
                serialize_alternative_form: false,
                port: Some(
                    8080,
                ),
                path: "/owner/repo",
            },
        ),
        tag: AtomTag(
            "foo",
        ),
        version: Some(
            VersionReq {
                comparators: [
                    Comparator {
                        op: Caret,
                        major: 1,
                        minor: None,
                        patch: None,
                        pre: Prerelease(""),
                    },
                ],
            },
        ),
    },
    Uri {
        url: Some(
            Url {
                scheme: Https,
                user: None,
                password: None,
                host: Some(
                    "foo.com",
                ),
                serialize_alternative_form: false,
                port: None,
                path: "/owner/repo",
            },
        ),
        tag: AtomTag(
            "bar",
        ),
        version: Some(
            VersionReq {
                comparators: [
                    Comparator {
                        op: Caret,
                        major: 1,
                        minor: None,
                        patch: None,
                        pre: Prerelease(""),
                    },
                ],
            },
        ),
    },
    Uri {
        url: Some(
            Url {
                scheme: Https,
                user: None,
                password: None,
                host: Some(
                    "github.com",
                ),
                serialize_alternative_form: false,
                port: None,
                path: "/owner/repo",
            },
        ),
        tag: AtomTag(
            "λ",
        ),
        version: Some(
            VersionReq {
                comparators: [
                    Comparator {
                        op: Caret,
                        major: 1,
                        minor: None,
                        patch: None,
                        pre: Prerelease(""),
                    },
                ],
            },
        ),
    },
    Uri {
        url: Some(
            Url {
                scheme: Https,
                user: None,
                password: None,
                host: Some(
                    "github.com",
                ),
                serialize_alternative_form: false,
                port: None,
                path: "/owner/repo",
            },
        ),
        tag: AtomTag(
            "λ",
        ),
        version: Some(
            VersionReq {
                comparators: [
                    Comparator {
                        op: Caret,
                        major: 1,
                        minor: None,
                        patch: None,
                        pre: Prerelease(""),
                    },
                ],
            },
        ),
    },
    Uri {
        url: Some(
            Url {
                scheme: Ssh,
                user: Some(
                    "git",
                ),
                password: None,
                host: Some(
                    "github.com",
                ),
                serialize_alternative_form: false,
                port: None,
                path: "/owner/repo",
            },
        ),
        tag: AtomTag(
            "λ",
        ),
        version: Some(
            VersionReq {
                comparators: [
                    Comparator {
                        op: Caret,
                        major: 1,
                        minor: None,
                        patch: None,
                        pre: Prerelease(""),
                    },
                ],
            },
        ),
    },
    Uri {
        url: Some(
            Url {
                scheme: Https,
                user: None,
                password: None,
                host: Some(
                    "github.com",
                ),
                serialize_alternative_form: false,
                port: None,
                path: "/nixos/nixpkgs",
            },
        ),
        tag: AtomTag(
            "zlib",
        ),
        version: Some(
            VersionReq {
                comparators: [
                    Comparator {
                        op: Caret,
                        major: 1,
                        minor: None,
                        patch: None,
                        pre: Prerelease(""),
                    },
                ],
            },
        ),
    },
    Uri {
        url: Some(
            Url {
                scheme: Https,
                user: Some(
                    "user",
                ),
                password: Some(
                    "password",
                ),
                host: Some(
                    "example.com",
                ),
                serialize_alternative_form: false,
                port: None,
                path: "/my/repo",
            },
        ),
        tag: AtomTag(
            "id",
        ),
        version: Some(
            VersionReq {
                comparators: [
                    Comparator {
                        op: Caret,
                        major: 0,
                        minor: Some(
                            2,
                        ),
                        patch: None,
                        pre: Prerelease(""),
                    },
                ],
            },
        ),
    },
    Uri {
        url: Some(
            Url {
                scheme: Https,
                user: Some(
                    "user",
                ),
                password: Some(
                    "password",
                ),
                host: Some(
                    "example.com",
                ),
                serialize_alternative_form: false,
                port: None,
                path: "/my/repo",
            },
        ),
        tag: AtomTag(
            "id",
        ),
        version: Some(
            VersionReq {
                comparators: [
                    Comparator {
                        op: Caret,
                        major: 0,
                        minor: Some(
                            2,
                        ),
                        patch: None,
                        pre: Prerelease(""),
                    },
                ],
            },
        ),
    },
    Uri {
        url: Some(
            Url {
                scheme: Ssh,
                user: Some(
                    "user",
                ),
                password: None,
                host: Some(
                    "example.com",
                ),
                serialize_alternative_form: false,
                port: None,
                path: "/my/repo",
            },
        ),
        tag: AtomTag(
            "id",
        ),
        version: Some(
            VersionReq {
                comparators: [
                    Comparator {
                        op: Caret,
                        major: 0,
                        minor: Some(
                            2,
                        ),
                        patch: None,
                        pre: Prerelease(""),
                    },
                ],
            },
        ),
    },
    Uri {
        url: Some(
            Url {
                scheme: Https,
                user: None,
                password: None,
                host: Some(
                    "github.com",
                ),
                serialize_alternative_form: false,
                port: None,
                path: "/owner/repo",
            },
        ),
        tag: AtomTag(
            "yep",
        ),
        version: Some(
            VersionReq {
                comparators: [
                    Comparator {
                        op: Caret,
                        major: 1,
                        minor: None,
                        patch: None,
                        pre: Prerelease(""),
                    },
                ],
            },
        ),
    },
    Uri {
        url: Some(
            Url {
                scheme: Https,
                user: None,
                password: None,
                host: Some(
                    "bitbucket.org",
                ),
                serialize_alternative_form: false,
                port: None,
                path: "/",
            },
        ),
        tag: AtomTag(
            "yep",
        ),
        version: Some(
            VersionReq {
                comparators: [
                    Comparator {
                        op: Caret,
                        major: 1,
                        minor: None,
                        patch: None,
                        pre: Prerelease(""),
                    },
                ],
            },
        ),
    },
    Uri {
        url: Some(
            Url {
                scheme: File,
                user: None,
                password: None,
                host: None,
                serialize_alternative_form: true,
                port: None,
                path: "foo/bar/baz",
            },
        ),
        tag: AtomTag(
            "my-atom",
        ),
        version: None,
    },
    Uri {
        url: Some(
            Url {
                scheme: File,
                user: None,
                password: None,
                host: None,
                serialize_alternative_form: true,
                port: None,
                path: "/foo/bar/baz",
            },
        ),
        tag: AtomTag(
            "my-atom",
        ),
        version: None,
    },
    Uri {
        url: Some(
            Url {
                scheme: Ssh,
                user: None,
                password: None,
                host: Some(
                    "my.ssh.com",
                ),
                serialize_alternative_form: false,
                port: None,
                path: "/my/repo",
            },
        ),
        tag: AtomTag(
            "hello",
        ),
        version: None,
    },
    Uri {
        url: None,
        tag: AtomTag(
            "foo",
        ),
        version: Some(
            VersionReq {
                comparators: [
                    Comparator {
                        op: Caret,
                        major: 0,
                        minor: Some(
                            8,
                        ),
                        patch: None,
                        pre: Prerelease(""),
                    },
                ],
            },
        ),
    },
    Uri {
        url: None,
        tag: AtomTag(
            "foo",
        ),
        version: None,
    },
    Uri {
        url: None,
        tag: AtomTag(
            "foo",
        ),
        version: None,
    },
]
